xray/segment.go

```diff
@@ -322,25 +354,27 @@ func (seg *Segment) flush() {
 			logger.Debugf("emit lambda subsegment named: %v", seg.Name)
 			seg.emit()
 		} else {
-			seg.parent.safeFlush()
+			return false
 		}
 	}
-}
-
-func (seg *Segment) safeFlush() {
-	seg.Lock()
-	defer seg.Unlock()
-	seg.openSegments--
-	seg.flush()
+	return true
 }
 
 func (seg *Segment) safeInProgress() bool {
-	seg.Lock()
+	seg.RLock()
 	b := seg.InProgress
-	seg.Unlock()
+	seg.RUnlock()
 	return b
 }
 
+// getName returns name of the segment. This method is thread safe.
+func (seg *Segment) getName() string {
```

Comment:
What is the purpose of this? I get that you can wrap every single getter call with `Rlock()/Runlock()` without changing the correctness of the program. We still need to use locks wisely.

Reply:
I explained the use case in the previous comment. Currently this is called only by `aws.go` closing `connect` subsegment logic.